rfcs/0160-portable-service-layer.md

```diff
@@ -0,0 +1,893 @@
+accepts a new parameter: `forceDisableUserChange`. When this property is
+enabled, the service configured in such a way that it does not change users,
+which an unprivileged user is not allowed to do. With this setting enabled, it
+becomes possible for an unprivileged user to deploy PostgreSQL.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+The `processes.nix` expression shown earlier can be used with a variety of tools
+to deploy running process instances.
+
+The following command deploys the configurations as `sysvinit` scripts and
+activates the configuration automatically:
+
+```bash
+$ nixproc-sysvinit-switch processes.nix
+```
+
+We can deploy the same configuration as systemd units with the following command:
+
+```bash
+$ nixproc-systemd-switch processes.nix
+```
+
+Or with `supervisord` with the following command:
+
+```bash
+$ nixproc-supervisord-switch processes.nix
+```
+
+As an unprivileged user, we can use the `--state-dir` parameter to change the
+prefix of the state directories to reside in a user's home directory
+(`/home/sander/var`) and disable user changing:
+
+```bash
+$ nixproc-sysvinit-switch --state-dir /home/sander/var --force-disable-user-change processes.nix
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The ideas describes in this RFC makes the Nixpkgs eco-system more powerful, but
+these new features come at at a price:
+
+* Using high-level abstractions hide the low-level configuration details of
+  specific process managers. This makes it sometimes more difficult to reason
+  about what is happening in problematic cases
+* More testing is required -- we need to test services in multiple deployment
+  scenarios, which is more expensive. Fortunately, I have seen that testing for
+  two kinds of targets often give enough certainty: if you test with a
+  foreground process (e.g. `systemd`) and a daemon (e.g. `sysvinit`) there is a
+  big chance that it will work with other process managers as well.
+* You may target the lowest common denominator. This is not actually true,
+  because with overrides it is still possible to use any feature of a process
+  manager. Moreover, not all services need to be portable -- low-level services
+  such as `cups` and `syslogd` can still be packaged in a process
+  manager-specific way.
+
+# Alternatives
+[alternatives]: #alternatives
+
+This RFC proposes a new strategy to allow multiple process instances of the
+same service to co-exist. Co-existence is possible by avoiding conflicts --
+making sure that state directories and port numbers do not conflict through
+function parameters.
+
+It is also possible to isolate services by using concepts such as Linux
+namespaces. Then it is not required to avoid certain kinds of conflicts.
+Unfortunately, using these features is restricted to Linux only -- this RFC aims
+to be operating system agnostic.
+
+# Prior art
+[prior-art]: #prior-art
```

Comment:
I'm surprised for you to not mention your own [Disnix](https://github.com/svanderburg/disnix)/[Dysnomia](https://github.com/svanderburg/dysnomia) projects. It's clear that that's where this RFC comes from.

Reply:
Looking for something like this, I wast pointed to https://devenv.sh/services/ as the most likely current candidate. I think it would make sense to add that to prior art as well.

Reply:
Or kubenix, or arion (which piggybacks on NixOS instead of rewriting the service library again). Maybe more?
You might want to draw the line. They're just custom service managers for which this RFC proposes to create a common interface. If someone has a repo for services on S6, would that change the design? Probably not.

Reply:
I've expanded this section a bit. The reason why I didn't add my own projects is because they are only loosely coupled to process management -- in its current state, whenever Disnix needs to manage processes it invokes the Nix process management framework for that purpose.

I have also added `devenv`.